src/Content/Text/BBCode.php

```diff
@@ -1256,643 +1256,626 @@ public static function convert($text, $try_oembed = true, $simple_html = self::I
 	{
 		$a = DI::app();
 
-		/*
-		 * preg_match_callback function to replace potential Oembed tags with Oembed content
-		 *
-		 * $match[0] = [tag]$url[/tag] or [tag=$url]$title[/tag]
-		 * $match[1] = $url
-		 * $match[2] = $title or absent
-		 */
-		$try_oembed_callback = function ($match)
-		{
-			$url = $match[1];
-			$title = $match[2] ?? null;
-
-			try {
-				$return = OEmbed::getHTML($url, $title);
-			} catch (Exception $ex) {
-				$return = $match[0];
-			}
-
-			return $return;
-		};
-
-		// Extracting code blocks before the whitespace processing and the autolinker
-		$codeblocks = [];
-
-		$text = preg_replace_callback("#\[code(?:=([^\]]*))?\](.*?)\[\/code\]#ism",
-			function ($matches) use (&$codeblocks) {
-				$return = '#codeblock-' . count($codeblocks) . '#';
-				if (strpos($matches[2], "\n") !== false) {
-					$codeblocks[] = '<pre><code class="language-' . trim($matches[1]) . '">' . htmlspecialchars(trim($matches[2], "\n\r"), ENT_NOQUOTES, 'UTF-8') . '</code></pre>';
-				} else {
-					$codeblocks[] = '<code>' . htmlspecialchars($matches[2], ENT_NOQUOTES, 'UTF-8') . '</code>';
-				}
-
-				return $return;
-			},
-			$text
-		);
-
-		// Hide all [noparse] contained bbtags by spacefying them
-		// POSSIBLE BUG --> Will the 'preg' functions crash if there's an embedded image?
+		$text = self::performWithEscapedTags($text, ['code'], function ($text) use ($try_oembed, $simple_html, $for_plaintext, $a) {
+			$text = self::performWithEscapedTags($text, ['noparse', 'nobb', 'pre'], function ($text) use ($try_oembed, $simple_html, $for_plaintext, $a) {
```

Comment:
Is this a function inside a function that's inside a function?

These functions inside of functions are nice for short functions where you can see the whole function at once. But here I'm losing track of what is inside the function and what is outside.

Reply:
That's exactly why I created these constructs, the indentation helps to remind of the escaped blocks. Otherwise in a flat method like it was before, you could forget that code blocks were escaped. But yeah  `BBCode::convert` isn't the most readable method even without the anonymous function callbacks.